eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala

```diff
@@ -146,421 +129,143 @@ class Peer(val nodeParams: NodeParams, remoteNodeId: PublicKey, authenticator: A
-          goto(DISCONNECTED) using DisconnectedData(d.address_opt, d.channels.collect { case (k: FinalChannelId, v) => (k, v) })
-        }
-
-      case Event(Terminated(actor), d: ConnectedData) if d.channels.values.toSet.contains(actor) =>
-        // we will have at most 2 ids: a TemporaryChannelId and a FinalChannelId
-        val channelIds = d.channels.filter(_._2 == actor).keys
-        log.info(s"channel closed: channelId=${channelIds.mkString("/")}")
-        if (d.channels.values.toSet - actor == Set.empty) {
-          log.info(s"that was the last open channel, closing the connection")
-          d.transport ! PoisonPill
-        }
-        stay using d.copy(channels = d.channels -- channelIds)
+    case Event(Disconnect(nodeId), d: ConnectedData) if nodeId == remoteNodeId =>
+      log.info("disconnecting")
+      sender ! "disconnecting"
+      d.peerConnection ! PoisonPill
+      stay
 
-      case Event(h: Authenticator.Authenticated, d: ConnectedData) =>
-        log.info(s"got new transport while already connected, switching to new transport")
-        context unwatch d.transport
-        d.transport ! PoisonPill
+    case Event(Terminated(actor), d: ConnectedData) if actor == d.peerConnection =>
+      Logs.withMdc(diagLog)(Logs.mdc(category_opt = Some(Logs.LogCategory.CONNECTION))) {
+        log.info("connection lost")
+      }
+      if (d.channels.isEmpty) {
+        // we have no existing channels, we can forget about this peer
+        stopPeer()
+      } else {
         d.channels.values.toSet[ActorRef].foreach(_ ! INPUT_DISCONNECTED) // we deduplicate with toSet because there might be two entries per channel (tmp id and final id)
-        self ! h
-        goto(DISCONNECTED) using DisconnectedData(d.address_opt, d.channels.collect { case (k: FinalChannelId, v) => (k, v) })
+        goto(DISCONNECTED) using DisconnectedData(d.channels.collect { case (k: FinalChannelId, v) => (k, v) })
+      }
 
-      case Event(unhandledMsg: LightningMessage, d: ConnectedData) =>
-        // we ack unhandled messages because we don't want to block further reads on the connection
-        d.transport ! TransportHandler.ReadAck(unhandledMsg)
-        log.warning(s"acking unhandled message $unhandledMsg")
-        stay
-    }
+    case Event(Terminated(actor), d: ConnectedData) if d.channels.values.toSet.contains(actor) =>
+      // we will have at most 2 ids: a TemporaryChannelId and a FinalChannelId
+      val channelIds = d.channels.filter(_._2 == actor).keys
+      log.info(s"channel closed: channelId=${channelIds.mkString("/")}")
+      if (d.channels.values.toSet - actor == Set.empty) {
+        log.info(s"that was the last open channel, closing the connection")
+        d.peerConnection ! PoisonPill
+      }
+      stay using d.copy(channels = d.channels -- channelIds)
+
+    case Event(connectionReady: PeerConnection.ConnectionReady, d: ConnectedData) =>
```

Comment:
Hah this is one is painful...once we go to DISCONNECTED the reconnectTask will also try to reconnect for no reason.
Do you think the randomization in the reconnectTask is enough, or should we explicitly tell the reconnectTask to NOT reconnect (which may lead to way too many particular cases to handle :/)?

Reply:
The randomization is there for precisely that purpose, not sure what we can do more. What if we refrain from reconnecting and our peer does the same?

Reply:
Yeah this is probably working in practice, it's just really hard to evaluate by just looking at the code...but we'll learn as we watch what happens on our node